include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,547 @@
+
+                __clear_wglocal_histograms(__hacc_private, __wgroup_idx * __num_bins, __num_bins, __self_item);
+
+                if (__seg_start + __work_group_size * __iters_per_work_item < __n)
+                {
+                    for (::std::size_t __idx = 0; __idx < __iters_per_work_item; __idx++)
+                    {
+                        ::std::size_t __val_idx = __seg_start + __idx * __work_group_size + __self_lidx;
+                        __accum_local_atomics_iter<_histogram_index_type, _atomic_address_space>(
+                            __input, __val_idx, __hacc_private, __wgroup_idx * __num_bins, __func);
+                    }
+                }
+                else
+                {
+                    for (::std::size_t __idx = 0; __idx < __iters_per_work_item; __idx++)
+                    {
+                        ::std::size_t __val_idx = __seg_start + __idx * __work_group_size + __self_lidx;
+                        if (__val_idx < __n)
+                        {
+                            __accum_local_atomics_iter<_histogram_index_type, _atomic_address_space>(
+                                __input, __val_idx, __hacc_private, __wgroup_idx * __num_bins, __func);
+                        }
+                    }
+                }
+
+                __dpl_sycl::__group_barrier(__self_item);
+
+                __reduce_out_histograms<_bin_type>(__hacc_private, __wgroup_idx * __num_bins, __bins, __num_bins,
+                                                   __self_item);
+            });
+    });
+}
+
+template <::std::uint16_t __iters_per_work_item, typename _ExecutionPolicy, typename _Iter1, typename _Iter2,
+          typename _Size, typename _IdxHashFunc, typename... _Range>
+inline auto
+__parallel_histogram_sycl_impl(_ExecutionPolicy&& exec, _Iter1 __first, _Iter1 __last, _Iter2 __histogram_first,
+                               const _Size& __num_bins, _IdxHashFunc __func, _Range&&... __opt_range)
+{
+    using __local_histogram_type = ::std::uint32_t;
+    using __global_histogram_type = typename ::std::iterator_traits<_Iter2>::value_type;
+
+    ::std::size_t __max_wgroup_size = oneapi::dpl::__internal::__max_work_group_size(exec);
+
+    ::std::uint16_t __work_group_size = ::std::min(::std::size_t(1024), __max_wgroup_size);
+
+    auto __local_mem_size = exec.queue().get_device().template get_info<sycl::info::device::local_mem_size>();
+    constexpr ::std::uint8_t __max_work_item_private_bins = 16;
+
+    auto keep_bins =
+        oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::write, _Iter2>();
+    auto bins_buf = keep_bins(__histogram_first, __histogram_first + __num_bins);
+
+    auto __f = oneapi::dpl::__internal::fill_functor<__global_histogram_type>{__global_histogram_type{0}};
+    //fill histogram bins with zeros
+    auto init_e = oneapi::dpl::__par_backend_hetero::__parallel_for(
```

Comment:
Since we are initializing to all zeros, could it potentially be faster to initialize all bytes as 0 with `queue::memset` instead of a parallel_for with a fill functor?

Reply:
This was my original approach in prototyping, but since the API has user supplied **iterator** input,  I think this may be the best we can do.  
I user could provide a `permutation_iterator` in for instance as the output histogram.